src/lib.rs

```diff
@@ -12,99 +12,155 @@
+
+        // Read the password.
+        let mut password = match get_reader().read_line() {
+            Ok(val) => { val },
+            Err(err) => {
+                // Reset the terminal and quit.
+                try!(termios::tcsetattr(STDIN_FILENO, termios::TCSANOW, &term_orig));
+
+                // Return the original IoError.
+                return Err(err);
+            }
+        };
+
+        // Reset the terminal and quit.
+        match termios::tcsetattr(STDIN_FILENO, termios::TCSANOW, &term_orig) {
+            Ok(_) => {},
+            Err(err) => {
+                unsafe { password.as_mut_vec() }.set_memory(0);
+                return Err(err);
+            }
+        }
 
-    // Read the password.
-    let mut password = match get_reader().read_line() {
-        Ok(val) => { val },
-        Err(err) => {
-            // Reset the terminal and quit.
-            try!(termios::tcsetattr(STDIN_FILENO, termios::TCSANOW, &term_orig));
+        // Remove the \n from the line.
+        password.pop().unwrap();
 
-            // Return the original IoError.
-            return Err(err);
-        }
-    };
-
-    // Reset the terminal and quit.
-    match termios::tcsetattr(STDIN_FILENO, termios::TCSANOW, &term_orig) {
-        Ok(_) => {},
-        Err(err) => {
-            unsafe { password.as_mut_vec() }.set_memory(0);
-            return Err(err);
-        }
+        Ok(password)
     }
 
-    // Remove the \n from the line.
-    password.pop().unwrap();
-
-    Ok(password)
+    #[test]
+    fn it_works() {
+        let term_before = termios::Termios::from_fd(STDIN_FILENO).unwrap();
+        assert_eq!(read_password().unwrap().as_slice(), "my-secret");
+        let term_after = termios::Termios::from_fd(STDIN_FILENO).unwrap();
+        assert_eq!(term_before, term_after);
+        unsafe { TEST_EOF = true; }
+        assert!(!read_password().is_ok());
+        let term_after = termios::Termios::from_fd(STDIN_FILENO).unwrap();
+        assert_eq!(term_before, term_after);
+        assert!(unsafe { TEST_HAS_SEEN_REGULAR_BUFFER });
+        assert!(unsafe { TEST_HAS_SEEN_EOF_BUFFER });
+    }
 }
-
-#[test]
-fn it_works() {
-    let term_before = termios::Termios::from_fd(STDIN_FILENO).unwrap();
-    assert_eq!(read_password().unwrap().as_slice(), "my-secret");
-    let term_after = termios::Termios::from_fd(STDIN_FILENO).unwrap();
-    assert_eq!(term_before, term_after);
-    unsafe { TEST_EOF = true; }
-    assert!(!read_password().is_ok());
-    let term_after = termios::Termios::from_fd(STDIN_FILENO).unwrap();
-    assert_eq!(term_before, term_after);
-    assert!(unsafe { TEST_HAS_SEEN_REGULAR_BUFFER });
-    assert!(unsafe { TEST_HAS_SEEN_EOF_BUFFER });
+#[cfg(windows)]
+mod windows {
+    extern crate winapi;
+    extern crate "kernel32-sys" as kernel32;
+    use std::old_io::{IoError, IoErrorKind, IoResult};
+    use std::ptr::null_mut;
+    pub fn read_password() -> IoResult<String> {
+        unsafe {
```

Comment:
I'd rather not have such a large `unsafe { ... }` block. Only stuff that is actually unsafe should be marked as such. What's the reason the whole function uses `unsafe` code?